src/kws.c

```diff
@@ -210,25 +211,50 @@ static int ws_client_handshake(kws_t *kws)
 {
 	unsigned char nonce[16] = { 0 };
 	unsigned char enonce[128] = { 0 };
-	char req[2048] = { 0 };
+	char req[12048] = { 0 };
 	char *frame_end = NULL;
+	char *extra_headers = NULL;
 
 	gen_nonce(nonce, sizeof(nonce));
 	b64encode(nonce, sizeof(nonce), enonce, sizeof(enonce));
 
+	if (kws->params) {
+		ks_json_t *headers = ks_json_get_object_item(kws->params, "headers");
+
+		if (headers) {
+			ks_json_t *param;
+
+			KS_JSON_ARRAY_FOREACH(param, headers) {
+				const char *k = ks_json_get_object_string(param, "key", NULL);
+				const char *v = ks_json_get_object_string(param, "value", NULL);
+
+				if (k && v) {
+					if (extra_headers) {
+						extra_headers = ks_mprintf("%s%s: %s\r\n", headers, k, v);
```

Comment:
I think this leaks memory.  Also- is it correct to pass in "headers" here?  I think you mean "extra_headers"?